### PR TITLE
Adding a DeleteOnRestartManager.

### PR DESCRIPTION
**Pull request by harshgMSFT**
Does not contain tests

### CODE REVIEW

**Comment by harshgMSFT** [comment]
@yishaigalatzer , @deepakaravindr @pranavkm @emgarten

File: src/PackageManagement.VisualStudio/IDE/EnvDTEProjectName.cs [hunk 1]
```diff
@@ -1,49 +1,49 @@
-﻿// Copyright (c) .NET Foundation. All rights reserved.
-// Licensed under the Apache License, Version 2.0. See License.txt in the project root for license information.
```
Comment (deepakaravindr) [inline src/PackageManagement.VisualStudio/IDE/EnvDTEProjectName.cs:2]: Can you please fix the line-ending diff that shows up in this file and all the other files? This shows up as a diff, even though it is unrelated to your fix

File: src/PackageManagement.VisualStudio/IDE/VSSolutionManager.cs [hunk 2]
```diff
@@ -61,6 +61,9 @@ public NuGetProject DefaultNuGetProject
         public event EventHandler<NuGetProjectEventArgs> NuGetProjectRemoved;
         public event EventHandler<NuGetProjectEventArgs> NuGetProjectRenamed;
 
+      //  [Import]
```
Comment (deepakaravindr) [inline src/PackageManagement.VisualStudio/IDE/VSSolutionManager.cs]: Delete?

File: src/PackageManagement.VisualStudio/IDE/VSSolutionManager.cs [hunk 3]
```diff
@@ -598,10 +602,11 @@ public int OnCmdUIContextChanged(uint dwCmdUICookie, int fActive)
             {
                 ThreadHelper.ThrowIfNotOnUIThread();
                 OnSolutionExistsAndFullyLoaded();
+
                 // We must call DeleteMarkedPackageDirectories outside of OnSolutionOpened, because OnSolutionOpened might be called in the constructor
                 // and DeleteOnRestartManager requires VsFileSystemProvider and RepositorySetings which both have dependencies on SolutionManager.
                 // In practice, this code gets executed even when a solution is opened directly during Visual Studio startup.
```
Comment (deepakaravindr) [inline src/PackageManagement.VisualStudio/IDE/VSSolutionManager.cs]: Delete??

File: src/PackageManagement.VisualStudio/IDE/EnvDTEProjectName.cs (hunk 1 above)
Comment (harshgMSFT) [inline src/PackageManagement.VisualStudio/IDE/EnvDTEProjectName.cs:2]: yeah that is my bad, please ignore the commit marked as temp, all these are in that commit.

File: src/PackageManagement.VisualStudio/IDE/VSSolutionManager.cs (hunk 3 above)
Comment (harshgMSFT) [inline src/PackageManagement.VisualStudio/IDE/VSSolutionManager.cs]: I am not sure of the impact of not doing it here. The comment can ofcourse be deleted.

**Comment by harshgMSFT** [comment]
ping.

File: src/PackageManagement.VisualStudio/IDE/VSDeleteOnRestartManager.cs [hunk 4]
```diff
@@ -0,0 +1,85 @@
+﻿// Copyright (c) .NET Foundation. All rights reserved.
+// Licensed under the Apache License, Version 2.0. See License.txt in the project root for license information.
+
+using System;
+using System.Collections.Generic;
+using System.ComponentModel.Composition;
+using NuGet.Configuration;
+using NuGet.Packaging.Core;
+using NuGet.ProjectManagement;
+namespace NuGet.PackageManagement.VisualStudio
```
Comment (deepakaravindr) [inline src/PackageManagement.VisualStudio/IDE/VSDeleteOnRestartManager.cs:16]: Need an extra line

**Comment by harshgMSFT** [comment]
@halter73

File: src/PackageManagement.VisualStudio/IDE/VSDeleteOnRestartManager.cs [hunk 5]
```diff
@@ -0,0 +1,52 @@
...
+        }
+
+        private void OnSolutionOpenedOrClosed(object sender, EventArgs e)
+        {
+            // This is a solution event. Should be on the UI thread
+            ThreadHelper.ThrowIfNotOnUIThread();
+
+            // We need to do the check even on Solution Closed because, let's say if the yellow Update bar
+            // is showing and the user closes the solution; in that case, we want to hide the Update bar.
+            DeleteMarkedPackageDirectories(SolutionManager.NuGetProjectContext);
```
Comment (deepakaravindr) [inline src/PackageManagement.VisualStudio/IDE/VSDeleteOnRestartManager.cs:230]: Just note that SolutionManager's NuGetProjectContext may be null and it should be handler. We may consider using the output window in this context

**Comment by deepakaravindr** [comment]
As discussed, there will be a new iteration on this one. I will check it out. But, in general, changes look good to me. Please make sure that the line ending issue is fixed when you send the next iteration. If there are any changes in one of those files with line ending issues, I have not looked at them

**Comment by harshgMSFT** [comment]
@deepakaravindr since this PR is already out I cant wish away the temp commit, the only way would be to start a new PR. let me know if you want me to do that, however you did the right thing, there are no changes in the files with line endings.

**Comment by harshgMSFT** [comment]
updated.

File: src/PackageManagement.PowerShellCmdlets/PackageActionBaseCommand.cs [hunk 6]
```diff
@@ -17,6 +19,13 @@ namespace NuGet.PackageManagement.PowerShellCmdlets
 {
     public class PackageActionBaseCommand : NuGetPowerShellBaseCommand
     {
+        private IDeleteOnRestartManager _deleteOnRestartManager;
+
+        public PackageActionBaseCommand(IDeleteOnRestartManager deleteOnRestartManager)
```
Comment (harshgMSFT) [inline src/PackageManagement.PowerShellCmdlets/PackageActionBaseCommand.cs]: lets use the servicelocator here.

File: src/PackageManagement.VisualStudio/IDE/VSDeleteOnRestartManager.cs [hunk 7]
```diff
@@ -0,0 +1,52 @@
...
+using NuGet.Packaging.Core;
+using NuGet.ProjectManagement;
+
+namespace NuGet.PackageManagement.VisualStudio
+{
+    /// <summary>
+    /// An <see cref="DeleteOnRestartManager"/> manger which is used for surfacing errors and UI in VS.
+    /// </summary>
+    [Export(typeof(IDeleteOnRestartManager))]
+    public class VsDeleteOnRestartManager : DeleteOnRestartManager
```
Comment (harshgMSFT) [inline src/PackageManagement.VisualStudio/IDE/VSDeleteOnRestartManager.cs]: move everthing here.

File: src/VsConsole/Console/PowerConsoleToolWindow.cs [hunk 8]
```diff
@@ -58,6 +63,13 @@ private PowerConsoleWindow PowerConsoleWindow
             get { return ComponentModel.GetService<IPowerConsoleWindow>() as PowerConsoleWindow; }
         }
 
+        private IVsShell4 VsShell
+        {
+            get {
+
+                return this.GetService<IVsShell4>(typeof(SVsShell)); }
```
Comment (harshgMSFT) [inline src/VsConsole/Console/PowerConsoleToolWindow.cs]: formatting

**Comment by harshgMSFT** [comment]
updated.

File: src/NuGet.VisualStudio.Implementation/Extensibility/VsPackageUninstaller.cs [hunk 9]
```diff
@@ -43,7 +49,12 @@ public void UninstallPackage(Project project, string packageId, bool removeDepen
 
             ThreadHelper.JoinableTaskFactory.Run(async delegate
```
Comment (deepakaravindr) [inline src/NuGet.VisualStudio.Implementation/Extensibility/VsPackageUninstaller.cs:50]: You may already know this, but, you have to rebase your branch with 3.0-rtm. For example, this particular line would be different on the 3.0-rtm branch

File: src/PackageManagement.PowerShellCmdlets/NuGetPowerShellBaseCommand.cs [hunk 10]
```diff
@@ -66,6 +67,7 @@ protected NuGetPowerShellBaseCommand()
             SourceControlManagerProvider = ServiceLocator.GetInstance<ISourceControlManagerProvider>();
             _commonOperations = ServiceLocator.GetInstance<ICommonOperations>();
             PackageRestoreManager = ServiceLocator.GetInstance<IPackageRestoreManager>();
+            _deleteOnRestartManager = ServiceLocator.GetInstance<IDeleteOnRestartManager>();
```
Comment (deepakaravindr) [inline src/PackageManagement.PowerShellCmdlets/NuGetPowerShellBaseCommand.cs:70]: Not sure how we missed this in the review. You already have an instance of this _deleteOnRestartManager on the base class for all cmdlets, which is, NuGetPowershellBaseCommand. Just make this Property protected or something and use it from UninstallPackageCommand and PackageActionBaseCommand class. Why do we have the same property on several classes?

File: src/PackageManagement.VisualStudio/IDE/VSDeleteOnRestartManager.cs [hunk 11]
```diff
@@ -0,0 +1,233 @@
...
+                if (SolutionManager.SolutionDirectory != null)
+                {
+                    _packagesFolderPath =
+                        PackagesFolderPathUtility.GetPackagesFolderPath(SolutionManager, Settings);
+                }
+
+                return _packagesFolderPath;
+            }
+
+            set
```
Comment (deepakaravindr) [inline src/PackageManagement.VisualStudio/IDE/VSDeleteOnRestartManager.cs:85]: I don't think this setter or the local '_packagesFolderPath' is valid anymore. I would definitely make this getter only property and do the following
-                if (SolutionManager.SolutionDirectory != null)
-                {
-                    return
-                        PackagesFolderPathUtility.GetPackagesFolderPath(SolutionManager, Settings);
-                }
  +
-                return null;

**Comment by deepakaravindr** [comment]
After incorporating the comments, :shipit: . Thanks, @harshgMSFT

File: src/PackageManagement.PowerShellCmdlets/NuGetPowerShellBaseCommand.cs (hunk 10 above)
Comment (harshgMSFT) [inline src/PackageManagement.PowerShellCmdlets/NuGetPowerShellBaseCommand.cs:70]: :+1: yeah this was added at a later point ..

File: src/NuGet.VisualStudio.Implementation/Extensibility/VsPackageUninstaller.cs (hunk 9 above)
Comment (harshgMSFT) [inline src/NuGet.VisualStudio.Implementation/Extensibility/VsPackageUninstaller.cs:50]: yup.. need to rebase.

**Comment by harshgMSFT** [comment]
a26c2d844e1fd69efacfd47567fc47b7244de2e3

**Pull request closed without merging**